ethcore/src/private_transactions/encryptor.rs

```diff
@@ -0,0 +1,300 @@
+}
+
+struct EncryptionSession {
+	key: Bytes,
+	end_time: Instant,
+}
+
+/// SecretStore-based encryption/decryption operations.
+pub struct SecretStoreEncryptor {
+	client: FetchClient,
+	base_url: Option<String>,
+	threshold: u32,
+	sessions: Mutex<HashMap<Address, EncryptionSession>>,
+}
+
+impl SecretStoreEncryptor {
+	/// Create new encryptor with empty parameters.
+	pub fn empty() -> Result<Self, EthcoreError> {
+		Ok(SecretStoreEncryptor {
+			client: FetchClient::new()
+				.map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("{}", e))))?,
+			base_url: None,
+			threshold: 0,
+			sessions: Mutex::new(HashMap::new()),
+		})
+	}
+
+	/// Create new encryptor
+	pub fn new(key_server_url: Option<String>, threshold: u32) -> Result<Self, EthcoreError> {
+		Ok(SecretStoreEncryptor {
+			client: FetchClient::new()
+				.map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("{}", e))))?,
+			base_url: key_server_url,
+			threshold: threshold,
+			sessions: Mutex::new(HashMap::new()),
+		})
+	}
+
+	/// Ask secret store for key && decrypt the key.
+	fn retrieve_key(
+		&self,
+		url_suffix: &str,
+		method: FetchMethod,
+		contract_address: &Address,
+		contract_address_signature: &Signature,
+		requester: &Address,
+		unlocked_accounts: Arc<AccountProvider>
+	) -> Result<Bytes, EthcoreError> {
+		// check if the key was already cached
+		if let Some(key) = self.obtained_key(contract_address) {
+			return Ok(key);
+		}
+		// key id in SS is H256 && we have H160 here => expand with assitional zeros
+		let mut contract_address_extended = Vec::with_capacity(32);
+		contract_address_extended.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+		contract_address_extended.extend_from_slice(&**contract_address);
+
+		let base_url = self.base_url.clone().ok_or_else(|| PrivateTransactionError::KeyServerNotSet)?;
+
+		// prepare request url
+		let url = format!("{}/{}/{}{}",
+				base_url,
+				contract_address_extended.to_hex(),
+				contract_address_signature,
+				url_suffix,
+			);
+
+		// send HTTP request
+		let mut response = self.client.fetch_with_abort(&url, method, Default::default()).wait()
+			.map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("{}", e))))?;
+
+		if response.is_not_found() {
+			return Err(EthcoreError::PrivateTransaction(PrivateTransactionError::EncryptionKeyNotFound(*contract_address)));
+		}
+
+		if !response.is_success() {
+			return Err(EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(response.status().canonical_reason().unwrap_or("unknown").into())));
+		}
+
+		// read HTTP response
+		let mut result = String::new();
+		response.read_to_string(&mut result)?;
+
+		// response is JSON string (which is, in turn, hex-encoded, encrypted Public)
+		let result_len = result.len();
+		if result_len == 0 || &result[0..1] != "\"" || &result[result_len - 1..result_len] != "\"" {
```

Comment:
Please no. If it's a JSON please use `serde_json` to deserialize it.

Reply:
@grbIzl I don't have a test env for this functionality, but looks like this should work:
```
let encrypted_bytes: ethjson::bytes::Bytes = result.parse().map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(e)))?;
```
Don't know - why I've added this initially. Could you please check if this works?